Optimize AWS4Signer

danielmarbach opened the pull request:
This PR heavily optimizes the HTTP header path on the AWS4Signer to reduce the number of allocations.

## Description

See screenshots

## Motivation and Context

Making sure the request path is less allocation heavy will drastically improve the throughput of the AWS SDK.

## Testing
<!--- Please describe in detail how you tested your changes -->
<!--- Include details of your testing environment, and the tests you ran to -->
<!--- see how your change affects other areas of the code, etc. -->

## Screenshots (if appropriate)

### CanonicalHeaderNames

<img width="1324" alt="CanonicalizeHeaderNames" src="https://github.com/aws/aws-sdk-net/assets/174258/e0877f6f-8f6c-423e-8e34-51fbbe35c0d8">

https://github.com/danielmarbach/aws-sdk-net-benchmarks/blob/main/CanonicalHeaderNames.cs

### CanonicalHeaders

<img width="1403" alt="CanonicalHeaders" src="https://github.com/aws/aws-sdk-net/assets/174258/31abb69c-31e7-4eeb-af41-efbc57f6ae7b">

https://github.com/danielmarbach/aws-sdk-net-benchmarks/blob/main/CanonicalHeaderNames.cs

### CompressSpace

<img width="1385" alt="CompressSpace" src="https://github.com/aws/aws-sdk-net/assets/174258/a1ec2466-1605-422d-8cef-291dbbfa09f2">

https://github.com/danielmarbach/aws-sdk-net-benchmarks/blob/main/CompressSpace.cs

### CanonicalHeaders with CompressSpace

<img width="1005" alt="CanonicalHeadersCompressSpace" src="https://github.com/aws/aws-sdk-net/assets/174258/a7aadf47-86f5-46bd-9512-73f162fce2a4">

https://github.com/danielmarbach/aws-sdk-net-benchmarks/blob/main/CanonicalHeaderNames.cs

Attention this was just a short run because things take ages otherwise

## Types of changes
<!--- What types of changes does your code introduce? Put an `x` in all the boxes that apply: -->
- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist
<!--- Go over all the following points, and put an `x` in all the boxes that apply -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
- [ ] My code follows the code style of this project
- [ ] My change requires a change to the documentation
- [ ] I have updated the documentation accordingly
- [ ] I have read the **README** document
- [x] I have added tests to cover my changes
- [x] All new and existing tests passed

## License
<!--- The SDK is released under the [Apache 2.0 license][license], so any code you submit will be released under that license -->
<!--- For substantial contributions, we may ask you to sign a [Contributor License Agreement (CLA)][cla] -->
<!--- Put an `x` in the below box if you confirm that this request can be released under the Apache 2 license -->
- [ ] I confirm that this pull request can be released under the Apache 2 license

[issues]: https://github.com/aws/aws-sdk-net/issues
[license]: http://aws.amazon.com/apache2.0/
[cla]: http://en.wikipedia.org/wiki/Contributor_License_Agreement

danielmarbach commented:
@normj I would appreciate your input ;)

normj commented on sdk/src/Core/Amazon.Runtime/Internal/Util/ValueStringBuilder.cs:
```diff
@@ -0,0 +1,309 @@
+
+#nullable enable
+
+namespace AWSSDK.Core.NetStandard.Amazon.Runtime.Internal.Util
```
Since this is source coming from the `dotnet/runtime` repo lets move this to a folder called `RuntimeBackports` under the [ThirdParty](https://github.com/aws/aws-sdk-net/tree/main/sdk/src/Core/ThirdParty) namespace of core. The file's namespace should be `ThirdParty.RuntimeBackports`. I'll figure out the attribution.

normj commented on sdk/src/Core/Amazon.Runtime/Internal/Auth/AWS4Signer.cs:
```diff
@@ -886,6 +887,30 @@ protected internal static string CanonicalizeHeaders(IEnumerable<KeyValuePair<st
+        /// <param name="sortedHeaders">All request headers, sorted into canonical order</param>
+        /// <param name="builder">The value string builder to append to.</param>
+        /// <returns>Canonicalized string of headers, with the header names in lower case.</returns>
+        protected internal static void AppendCanonicalizedHeaders(ICollection<KeyValuePair<string, string>> sortedHeaders, ref ValueStringBuilder builder)
```
I'll need to do some research on why this was made protected but we don't want to expose `ValueStringBuilder` as part of the external API.

danielmarbach replied:
Since I haven't received more feedback I'm going with a lighter variant that will use ValueStringBuilder internally within the methods only. The impact should already be quite high and this will leave the API surface untouched for now.

Dreamescaper commented on sdk/src/Core/Amazon.Runtime/Internal/Auth/AWS4Signer.cs:
```diff
@@ -818,26 +819,26 @@ public static string DetermineService(IClientConfig clientConfig, IRequest reque
-            canonicalRequest.AppendFormat("{0}\n", CanonicalizeHeaders(sortedHeaders));
-            canonicalRequest.AppendFormat("{0}\n", CanonicalizeHeaderNames(sortedHeaders));
+            var canonicalRequestBuilder = new ValueStringBuilder(512);
+            canonicalRequestBuilder.Append($"{httpMethod}\n");
```
Does it make sense to replace such cases with
```csharp
            canonicalRequestBuilder.Append(httpMethod);
            canonicalRequestBuilder.Append('\n');
```

To avoid allocating new strings with '\n' in the end?

danielmarbach replied:
I would argue given the limited number of http methods the entire span could probably be optimized further
As far as I recall I could pull off a compiler trick here. I haven't gone as far as optimizing it to that level yet because there are still some outstanding inputs from the team about why some of these methods are even protected and whether the need to continue to be so

danielmarbach replied:
Addressed

danielmarbach commented:
@normj I have to update the description of the PR but I'm running out of my fun coding time allotted budget for now. I have added some benchmark results that should speak though themselves. Of course ideally it would be better if we could bring the ValueStringBuilder ref struct into the helper methods but since I did not get feedback on the usage of those protected static methods I went ahead and used an approach that doesn't require signature changes which I think is in your favor anyway for now.

@Dreamescaper I also addressed one of your feedbacks

danielmarbach commented on sdk/src/Core/Amazon.Runtime/Internal/Auth/AWS4Signer.cs:
```diff
@@ -893,16 +898,7 @@ protected internal static string CanonicalizeHeaders(IEnumerable<KeyValuePair<st
-            }
-            
-            return builder.ToString();
+            return string.Join(";", sortedHeaders.Select(x => x.Key.ToLowerInvariant()));
```
Ran another benchmark on Fedora and decided to switch to ValueStringBuilder here too

![image](https://github.com/aws/aws-sdk-net/assets/174258/9b578759-1376-44de-856a-110c4b1a3cdf)

peterrsongg reviewed (approved):
I ran the core and custom unit tests, mainly was concerned about the existing Signer tests which pass. Change looks good to me.

normj reviewed (approved):
Thanks @danielmarbach!
